### ISSUE TITLE
Custom nodes move to active flow on refresh

### ISSUE THREAD

**Issue by GerwinvBeek** [node-red/node-red#3005]
### What are the steps to reproduce?
Create two flows
Add default node red nodes + custom nodes to both flows
refresh page (ctrl + shift + r)

### What happens?
All custom nodes move to the active flow

### What do you expect to happen?
Nodes stay in the right flow

### Please tell us about your environment:

- [x] Node-RED version: 1.3.5
- [x] Node.js version: v12.18.4
- [x] npm version: 7.11.1
- [x] Platform/OS: windows
- [x] Browser: chrome
- [x] running in Docker: no

https://user-images.githubusercontent.com/26113706/120770962-00064100-c51f-11eb-81e1-4d4a8ff55485.mp4

[flows.txt](https://github.com/node-red/node-red/files/6596891/flows.txt)

**Comment by dceejay** [comment]
You can see the red deploy button - means you haven't saved the changes to the server... so when you refresh the page it reverts to the version the server knows about - not the one in your browser window.

**Comment by GerwinvBeek** [comment]
Hi @dceejay as you can see in the video, i click the deploy button before refreshing my page

**Comment by knolleary** [comment]
@dceejay they do click the deploy button prior to the refresh.

@GerwinvBeek you have shared a flows.txt file - what does it contain in relation to this issue? Is it the flows after the refresh? before the refresh?

Can you share any specific details about the custom nodes? Given we don't see this with any other nodes, there may be something particular about your nodes that is causing this. But without more information about them - such as a link to their code, it will be hard to make any progress on this.

**Comment by hardillb** [comment]
Can you reproduce this without your custom nodes?

What do those nodes do?

Is there any output in the node-red logs after the deploy?

**Comment by GerwinvBeek** [comment]
@knolleary, the linked txt are the flows after deploy (refresh doesn't change the json as far as i'm aware).
Here is a node that is causing problems,
The nodes are developed in typescript and compiled to javascript using gulp.
 
[CustomNode.zip](https://github.com/node-red/node-red/files/6597096/CustomNode.zip)

**Comment by knolleary** [comment]
@GerwinvBeek unfortunately that isn't a complete node module so I can't run with those nodes installed (its missing 'require.js' etc).

Looking the generated code, I can't see anything obvious about what could be causing this. So we really need a complete copy of your module so we can try to recreate it locally.

**Comment by GerwinvBeek** [comment]
@knolleary sorry about that,
here is the full folder:
[dist.zip](https://github.com/node-red/node-red/files/6597313/dist.zip)

**Comment by knolleary** [comment]
Hi @GerwinvBeek 

I still cannot get your nodes to load.

Your node html contains:
```
<script src="../require.js"></script>
```

That will not work because nothing is configured to serve up `require.js` on that path. Have you got some custom Node-RED settings to enable that?

**Comment by GerwinvBeek** [comment]
Hey @knolleary,

Sorry about the delay,
The custom node-red setting that I have enabled is:
```
    httpStatic: 'C:/Users/gbe/.node-red/node_modules/optimizers-warpspeed-ui-node-red/',
```
where the httpStatic refers to the folder where the nodes are installed.
I hope this helps!

**Comment by McTristan** [comment]
We have the same problem with our custom nodes. Whenever we refresh the browser after deploy all custom nodes are on the current flow tab and all flows are broken. We never had the problem with 1.2.x and 1.3.x but 2.0.4 and 2.0.5 breaks it.

**Comment by hardillb** [comment]
@McTristan Can you supply a copy of your custom node, this is most likely being driven by something in your node and the more examples we have the easier it will be to work out what it is.

**Comment by McTristan** [comment]
@hardillb Send you an E-Mail with the package and a screenshot

**Comment by McTristan** [comment]
Even when we delete all flows and subflows and import the (from version 1.2.9 exported) flows again. It looks fine after deploy but when we load the page again everything is messed up. Worse even - when we export these flows they are also messed up in the exported flows.json file.

**Comment by hardillb** [comment]
@McTristan Why does your node include node-red (v1.2.5) and @nodered/editor-client (v1.2.6) as runtime dependencies.

Node's shouldn't need these (especially not node-red it's self)

**Comment by hardillb** [comment]
@McTristan Also the package you have supplied doesn't even load (with Node-RED 2.0.5 and NodeJS 12). If your node requires NodeJS 14 then it should include a suitable `engine` clause in the package.json

```
2 Aug 09:33:41 - [warn] [@swpsoftware/node-red-contrib-swp-nodes/vicrm-api-remote-server] SyntaxError: Cannot use import statement outside a module (line:1)
2 Aug 09:33:41 - [warn] [@swpsoftware/node-red-contrib-swp-nodes/vicrm-api-remote-send-email] SyntaxError: Cannot use import statement outside a module (line:10)
2 Aug 09:33:41 - [warn] [@swpsoftware/node-red-contrib-swp-nodes/vicrm-api-remote-copy-email] SyntaxError: Cannot use import statement outside a module (line:10)
2 Aug 09:33:41 - [warn] [@swpsoftware/node-red-contrib-swp-nodes/vicrm-on-event] SyntaxError: Cannot use import statement outside a module (line:1)
2 Aug 09:33:41 - [warn] [@swpsoftware/node-red-contrib-swp-nodes/vicrm-api-remote-get-list] SyntaxError: Cannot use import statement outside a module (line:10)
2 Aug 09:33:41 - [warn] [@swpsoftware/node-red-contrib-swp-nodes/vicrm-api-remote-http-request-prepare] SyntaxError: Cannot use import statement outside a module (line:10)

```

**Comment by McTristan** [comment]
We are investigating the issue right now, however we are experiencing timing issues in our tests as the constructor of each node is not immediately called after registerType. Using the getNode method of the node-red-node-test-helper package, the result is always null unless we wait a few moments so it looks like there is a change in Node Red 2.0 when initializing the nodes. We have debugged it over and over and it seems to be a timing or rather call order issue.

We are not sure whom to ask for help - the maintainer of the node-red-node-test-helper package or you as we don't yet know what the underlying cause is.

Our tests look like this:

`describe('remote server node', () => {
    beforeEach((done: () => void) => {
        helper.start(done);
    });

    afterEach(async (done: () => void) => {
        await helper.stop(done);
    });

    test('should be loaded', async (done: () => void) => {
        const flow: TestFlowsItem[] = [FakeRemoteServerDefinition];

        await helper.load(remoteServerInit, flow, {});
        const nodeToTest = helper.getNode<RemoteServerNode>('serverConfig'); //thats were the node is null with nodered 2.x -> if we wait between helper.load and helper.getNode it works
        
        assert.that(nodeToTest.name).is.equalTo('testServer');
        done();
    });
});
`

**Comment by knolleary** [comment]
@McTristan Ben forwarded me a copy of your nodes so I could take a look.

Given we don't support nodes packaged as ES6 modules, how are you loading these nodes in the first place?

**Comment by McTristan** [comment]
Via the node Option -r esm , as we use a custom build docker container we also install the esm package automatically.

we build the image like this (gitlab pipeline):

`    - export NODEREDID=$(docker ps -aqf "name=nodered")
    - docker run -d --name nodered $IMAGE_TAG_NODERED
    - docker exec nodered npm install esm
    - docker commit nodered $IMAGE_TAG
    - docker stop nodered || echo "already stopped"
    - docker rm nodered || echo "already removed"
    - docker run -d --name nodered $IMAGE_TAG -e NODE_OPTIONS='-r esm'
    - docker cp . nodered:/noderednodes
    - docker exec nodered npm install esm
    - docker exec nodered npm install node-red-dashboard
    - docker exec nodered npm install node-red-contrib-aedes
    - docker exec nodered npm install /noderednodes
    - docker commit nodered $IMAGE_TAG
    - docker push $IMAGE_TAG
`

and use it like this (docker compose):

`nodered:    
    image: custom-nodered:latest
    container_name: nodered
    restart: unless-stopped
    ports:
      - "1880:1880"
    environment:
      - TZ
      - "NODE_OPTIONS=-r esm"
    volumes:
      - node-red-data:/data
`

in general we use typescript to develop our nodes so a compiler is also envolved.

**Comment by McTristan** [comment]
Github is removing the linefeeds but i think you can see what we do …

**Comment by hardillb** [comment]
If you edit the previous update and use triple back ticks (rather than single) on the line before and after the content it will preserve the formatting

**Comment by McTristan** [comment]
Done. Thank you!

**Comment by knolleary** [comment]
@McTristan the issue here is because you are use modules to load the node definitions in the editor, so they are loading asynchronously. We do have code to detect this and delay loading flows until everything is ready, but it isn't working for your nodes.

The quick fix for you is to change the module path to not begin with `./`. In other words, change
`src="./noderednodes/nodes/vicrm-api-remote-get-list-ui.js"` to `src="noderednodes/nodes/vicrm-api-remote-get-list-ui.js"` (and likewise in your other nodes).

Whilst you are at it, I would strongly recommend you change `noderednodes` to be something much more specific to your nodes/scope/org.

Depending on what version of Node-RED you want to support, we also added in 1.3 the ability for nodes to include static resources that are automatically made available to the editor without you having to create httpAdmin end points to do it - see https://nodered.org/docs/creating-nodes/resources for details.

**Comment by Expaso** [comment]
Bump :)

**Comment by knolleary** [comment]
Hi @Expaso

Do you have anything to add to this issue? Perhaps you are seeing similar behaviour with your custom nodes or someone else's?

If you are seeing this behaviour with a set of nodes other than those discussed so far, it would be help to share those details.

I have described the fix needed in the victron nodes. Does that apply in your case?

**Comment by McTristan** [comment]
@knolleary 
After investigating the problem quite some time it looks like we solved the problem on our site mainly by using the correct include path but we also got rid of the node-red dependencies and marked them as dev dependencies (which eslint did not like but be it that way). So on our side it works with the latest node red version.

We are still facing problems supporting Mqtt v5 or it seems to be broken somehow but this is another matter which we need to look into another time.

Thank you for your support!
